src/Language/Fixpoint/Solver/PLE.hs

```diff
@@ -969,6 +975,23 @@ evalApp γ ctx e0 es et
         else do
           return (Nothing, noExpand)
 
+evalApp _ ctx e0 es _
+  | EVar f <- dropECst e0
+  , Just rw <- lookupRewrite f $ icLRWs ctx
+  = do
+      -- expandedTerm <- elaborateExpr "EvalApp rewrite local:" $ eApps rw es
+      let expandedTerm = eApps rw es
+      modify $ \st -> st 
+        { evNewEqualities = S.insert (eApps e0 es, expandedTerm) (evNewEqualities st) }
+      return (Just expandedTerm, expand)
+
+evalApp _ ctx e0 es _
+  | deANFed <- deANF ctx e0
+  , dropECst deANFed /= dropECst e0
+  = do
+      return (Just $ eApps deANFed es, expand)
```

Comment:
We would need an example of an expression that requires to inline anf bindings to motivate this case.

Reply:
The issue is that I don't have a small example but the SKI example without lemmas doesn't pass without unfolding ANFs

Reply:
That sounds like we will need to do some debugging to understand why deANF is needed then.

Reply:
I suspect that is some bug in the code that does the deANFing because I suppose that ANF expressions shouldn't reach the evaluation loop

Reply:
Yes, REST does need to remove ANFs, and that use I think I understand.

Reply:
The weird thing is that the anf terms aren't introduced by the local rewrites they are already there, maybe is just the SMT solver that has trouble to see qualities when they are inside a lambda

Reply:
I'm guessing ANF has the potential to interfere with beta reduction, and also with unfoldings.

For instance, `anf1 eta1` won't reduce, unless PLE substitutes `anf1` to get `(\x -> e) eta1`.

In a similar way, PLE could get stuck in `anf2 y` if PLE needs to substitute `anf2` to get a saturated application `reflectedFun x y`. I think this situation is not a problem when we don't want to reason with partial applications, but with partial applications, a term like
```Haskell
let anf2 = (\eta1 -> reflectedFun x eta1)
 in f anf2
```
could have `f` unfolded as follows
```Haskell
let anf2 = (\eta1 -> reflectedFun x eta1)
 in anf2 y
```
and now we would want `anf2` to be inlined so PLE learns that `eta1` is `y` which might allow additional unfoldings.

Some problems of the current implementation:
1. The above issues can happen with any let bindings in the code, not only ANF bindings.
2. `deANF` is applied over and over on multiple subexpressions and the bindings they use, redoing the inlining in the bindings, and potentially introducing more subexpressions, on which deANF would be applied again. It seems unnecessarily wasteful when the same work can be done by a single pass on the involved bindings. Granted, other issues might be uncovered when traying the deANF only once.
3. `deANF` seems to be inlining even when the bindings are not ANF bindings. Adding traces shows for instance that `VV##F##63` is rewritten to `SKIDC.CVar σ##a1wT γ##a1wU x##a1x2`. If only ANF bindings should be inlined, this would need to be fixed.

I don't have a proposal to fix (1) and (2), but I think it is at least worth documenting them in comments, probably users would care of them as well.

Reply:
1 can be solved with local rewrites maybe
For 2 no idea

Reply:
Also we could compile lets to lambdas

Reply:
I contributed a fix for (3) in https://github.com/ucsd-progsys/liquid-fixpoint/pull/711

Reply:
SKIDC and the proposal are still passing? I would expect yes since it's a value

Reply:
Your tests continue to pass with it.

Reply:
I contributed another implementation that calls `deANF` only once per fork in the trie. It is passing all the tests as well. 2b8411b

In addition, it also explains the situations we have identified so far where ANF bindings interfere with PLE.

Reply:
Yeah looks way better